app/models/project_step.rb

```diff
@@ -135,6 +138,27 @@ def handle_original_date_logic
     end
   end
 
+  def handle_finalized_at
+    if is_finalized && ! finalized_at
+      self.finalized_at = Time.now
+    elsif ! is_finalized && finalized_at
+      self.finalized_at = nil
+    end
+  end
+
+  # Validates that a step may not be unfinalized more than 24 hours since it was previously marked
+  # as finalized.  Note, should generally be avoided by front-end logic, but guards against edge
+  # cases.
+  def unfinalize_allowed
+    if is_finalized_changed? && ! is_finalized && is_finalized_locked?
+      errors.add(:is_finalized, I18n.t("project_step.unfinalize_disallowed"))
+    end
+  end
```

Comment:
Yeah.  I've added tests for the finalize handling as well as the automatic original_date updates.  (and I got tripped up yet one more time with the confusing `original_date` getter which masks an empty db value)